File: examples/simple/Cpp/writer.cpp
#include "adios2.h"
#include <mpi.h>
#include <cstdlib>

int main(int argc, char **argv)
{
    const std::size_t nelems = 10;
	int KnownInt, RandomInt, KnownInts[nelems], RandomInts[nelems];
	int GlobalDims[1], Offsets[1], LocalDims[1];

	MPI_Comm comm;
	int rank, nproc;
	double dt=0.1;
	adios2::Variable<int> vKnownInt;
	adios2::Variable<int> vRandomInt;

	MPI_Init(&argc, &argv);
	MPI_Comm_dup(MPI_COMM_WORLD, &comm);
	MPI_Comm_rank(comm, &rank);
	MPI_Comm_size(comm, &nproc);


    //@effis-init comm=comm
	adios2::ADIOS adios(comm);

    if (rank == 0)
	{
        KnownInt = 666;
        RandomInt = rand() % 1000;
	}

	for (int i=0; i<nelems; i++)
	{
		KnownInts[i] = i;
	}

    //@effis-begin "Jabberwocky"->"Jaberwocky"
	adios2::IO io = adios.DeclareIO("Jabberwocky");

    GlobalDims[0] = nproc * nelems;
    Offsets[0] = rank * nelems;
    LocalDims[0] = nelems;
	adios2::Variable<int> vKnownInts  = io.DefineVariable<int>("KnownInts",  {nproc*nelems}, {rank*nelems}, {nelems}, adios2::ConstantDims);
    adios2::Variable<int> vRandomInts = io.DefineVariable<int>("RandomInts", {nproc*nelems}, {rank*nelems}, {nelems}, adios2::ConstantDims);
    if (rank == 0)
	{
		vKnownInt  = io.DefineVariable<int>("KnownInt");
        vRandomInt = io.DefineVariable<int>("RandomInt");
	}

	adios2::Engine engine = io.Open("Jabberwocky.bp", adios2::Mode::Write, comm);

    if (rank == 0)
	{
        engine.Put(vKnownInt,  KnownInt);
        engine.Put(vRandomInt, RandomInt);
	}

    for (int j=0; j<10; j++)
	{
        //@effis-timer start="LoopTimer", comm=comm
		
		for (int i=0; i<nelems; i++)
		{
			RandomInts[i] = rand() % 1000;
		}
        //@effis-timestep physical=j*dt, number=j
        engine.BeginStep();
        engine.Put(vKnownInts,  KnownInts);
        engine.Put(vRandomInts, RandomInts);
        engine.EndStep();

        //@effis-timer stop="LoopTimer"
	}

    engine.Close();
    //@effis-end

    //@effis-finalize
}

File: examples/simple/Python/writer.py
#!/usr/bin/env python

import adios2
import random
import numpy as np
import time
import sys

class mpi:
    UseComm = True

if mpi.UseComm:
    from mpi4py import MPI


if __name__ == "__main__":
    nelems = 10

    if mpi.UseComm:
        comm = MPI.COMM_WORLD
        adios = adios2.ADIOS(comm)
        rank = comm.Get_rank()
        size = comm.Get_size()
    else:
        adios = adios2.ADIOS()
        rank = 0
        size = 1
        comm = None


    #@effis-init comm=comm

    # Test scalars
    if rank == 0:
        KnownInt = np.array([666])
        RandomInt = np.random.randint(0, 1000, size=1)

    # Test arrays
    KnownInts = np.arange(0, nelems, dtype=np.int64)
    RandomInts = np.random.randint(0, 1000, size=nelems, dtype=np.int64)
    RandomInts2 = np.random.randint(0, 1000, size=nelems, dtype=np.int32)


    #@effis-begin engine--->"Jabberwocky"
    io = adios.DeclareIO("Jabberwocky")

    GlobalDims = [size * nelems]
    Offsets = [rank * nelems]
    LocalDims = [nelems]
    vKnownInts  = io.DefineVariable("KnownInts",  KnownInts,  GlobalDims, Offsets, LocalDims)
    vRandomInts = io.DefineVariable("RandomInts", RandomInts, GlobalDims, Offsets, LocalDims)
    vRandomInts2 = io.DefineVariable("RandomInts2", RandomInts2, GlobalDims, Offsets, LocalDims)
    if rank == 0:
        io.DefineVariable("KnownInt",  KnownInt,  [], [], [])
        vRandomInt = io.DefineVariable("RandomInt", RandomInt, [], [], [])


    # Open
    if mpi.UseComm:
        engine = io.Open("Jabberwocky.bp", adios2.Mode.Write, comm)
    else:
        engine = io.Open("Jabberwocky.bp", adios2.Mode.Write)


    # Write
    if rank == 0:
        vKnownInt = io.InquireVariable("KnownInt")
        engine.Put(vKnownInt,  KnownInt)
        engine.Put(vRandomInt, RandomInt)


    for i in range(10):
        #@effis-timer start="LoopTimer", comm=comm

        RandomInts = np.random.randint(0, 1000, size=nelems, dtype=np.int64)
        RandomInts2 = np.random.randint(0, 1000, size=nelems, dtype=np.int64)
        #@effis-timestep physical=i*0.01, number=i

        if i % 5 == 0:
            engine.BeginStep()
            engine.Put(vKnownInts,  KnownInts)
            engine.Put(vRandomInts, RandomInts)
            engine.Put(vRandomInts2, RandomInts2)
            engine.EndStep()

        time.sleep(1)

        #@effis-timer stop="LoopTimer"

    engine.Close()
    #@effis-end


    #@effis-finalize

